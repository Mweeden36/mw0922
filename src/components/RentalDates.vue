<template>
  <div class="rental-dates__wrapper" :class="invalid ? 'invalid' : ''">
    <b class="checkout-form__label"> Rental Dates: </b>
    <Datepicker
      class="date-picker start"
      v-model="rentalStart"
      :format="format"
      :clearable="false"
    />
    <span class="date-separator">to</span>
    <Datepicker
      class="date-picker end"
      v-model="rentalEnd"
      :format="format"
      :clearable="false"
    />
    <div v-if="invalid" class="error-notification color--danger">
      Please make sure the return date is after the checkout date.
    </div>
  </div>
</template>

<script>
import Datepicker from "@vuepic/vue-datepicker";

export default {
  name: "RentalDates",

  components: {
    Datepicker,
  },

  data() {
    const startDate = new Date();
    const defaultDaysToRent = 7;
    return {
      format: "MM/dd/yy",
      millisecondsInDay: 1000 * 60 * 60 * 24,
      rentalStart: startDate,
      rentalEnd: new Date(
        new Date().setDate(startDate.getDate() + defaultDaysToRent)
      ),
    };
  },

  computed: {
    daysToRent() {
      const diff = this.rentalEnd.getTime() - this.rentalStart.getTime();
      return Math.round(diff / this.millisecondsInDay);
    },

    invalid() {
      return this.daysToRent < 1;
    },
  },

  watch: {
    daysToRent() {
      this.$emit("input", {
        rentalStart: this.rentalStart,
        rentalEnd: this.rentalEnd,
        daysToRent: this.daysToRent,
      });
    },
  },

  mounted() {
    this.$emit("input", {
      rentalStart: this.rentalStart,
      rentalEnd: this.rentalEnd,
      daysToRent: this.daysToRent,
    });
  },
};
</script>

<style lang="scss" scoped>
.rental-dates__wrapper {
  &.invalid {
    :deep(.dp__input) {
      border-color: red;
    }
  }
  .date-picker {
    width: 150px;
    display: inline-block;
  }
  .date-separator {
    padding: 0px 5px;
  }
}
</style>

<style lang="scss">
// Can't use scoped styles because datepicker renders to body.
.dp__instance_calendar {
  .dp__button {
    display: none;
  }
}
</style>
