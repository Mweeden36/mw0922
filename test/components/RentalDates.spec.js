import RentalDates from "../../src/components/RentalDates.vue";
import { describe, it, expect } from "vitest";
import { mount } from "@vue/test-utils";

describe("RentalDates.vue", () => {
  it("should initialize with sensible defaults", () => {
    const { vm } = mount(RentalDates);
    expect(vm.daysToRent).toEqual(7);
    expect(vm.invalid).toEqual(false);
  });

  it("should determine when the dates are too close together", () => {
    const { vm } = mount(RentalDates);
    vm.rentalStart = new Date();
    vm.rentalEnd = new Date();
    expect(vm.invalid).toEqual(true);
  });
});
