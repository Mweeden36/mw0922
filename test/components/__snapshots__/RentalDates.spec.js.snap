// Vitest Snapshot v1

exports[`RentalDates.vue > should generate the correct html 1`] = `
"<div class=\\"rental-dates__wrapper\\" data-v-36d48911=\\"\\"><b class=\\"checkout-form__label\\" data-v-36d48911=\\"\\"> Rental Dates: </b>
  <div class=\\"dp__main dp__theme_light date-picker start\\" data-v-36d48911=\\"\\">
    <div aria-label=\\"Datepicker input\\" role=\\"textbox\\" aria-multiline=\\"false\\" aria-disabled=\\"false\\" aria-readonly=\\"false\\">
      <!---->
      <div class=\\"dp__input_wrap\\">
        <!----><input class=\\"dp__pointer dp__input_readonly dp__input dp__input_icon_pad dp__input_reg\\" inputmode=\\"none\\" placeholder=\\"\\" autocomplete=\\"off\\">
        <!----><svg version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"32\\" height=\\"32\\" viewBox=\\"0 0 32 32\\" class=\\"dp__icon dp__input_icon dp__input_icons\\">
          <path d=\\"M29.333 8c0-2.208-1.792-4-4-4h-18.667c-2.208 0-4 1.792-4 4v18.667c0 2.208 1.792 4 4 4h18.667c2.208 0 4-1.792 4-4v-18.667zM26.667 8v18.667c0 0.736-0.597 1.333-1.333 1.333 0 0-18.667 0-18.667 0-0.736 0-1.333-0.597-1.333-1.333 0 0 0-18.667 0-18.667 0-0.736 0.597-1.333 1.333-1.333 0 0 18.667 0 18.667 0 0.736 0 1.333 0.597 1.333 1.333z\\"></path>
          <path d=\\"M20 2.667v5.333c0 0.736 0.597 1.333 1.333 1.333s1.333-0.597 1.333-1.333v-5.333c0-0.736-0.597-1.333-1.333-1.333s-1.333 0.597-1.333 1.333z\\"></path>
          <path d=\\"M9.333 2.667v5.333c0 0.736 0.597 1.333 1.333 1.333s1.333-0.597 1.333-1.333v-5.333c0-0.736-0.597-1.333-1.333-1.333s-1.333 0.597-1.333 1.333z\\"></path>
          <path d=\\"M4 14.667h24c0.736 0 1.333-0.597 1.333-1.333s-0.597-1.333-1.333-1.333h-24c-0.736 0-1.333 0.597-1.333 1.333s0.597 1.333 1.333 1.333z\\"></path>
        </svg>
        <!---->
        <!---->
      </div>
    </div>
    <!---->
  </div><span class=\\"date-separator\\" data-v-36d48911=\\"\\">to</span>
  <div class=\\"dp__main dp__theme_light date-picker end\\" data-v-36d48911=\\"\\">
    <div aria-label=\\"Datepicker input\\" role=\\"textbox\\" aria-multiline=\\"false\\" aria-disabled=\\"false\\" aria-readonly=\\"false\\">
      <!---->
      <div class=\\"dp__input_wrap\\">
        <!----><input class=\\"dp__pointer dp__input_readonly dp__input dp__input_icon_pad dp__input_reg\\" inputmode=\\"none\\" placeholder=\\"\\" autocomplete=\\"off\\">
        <!----><svg version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"32\\" height=\\"32\\" viewBox=\\"0 0 32 32\\" class=\\"dp__icon dp__input_icon dp__input_icons\\">
          <path d=\\"M29.333 8c0-2.208-1.792-4-4-4h-18.667c-2.208 0-4 1.792-4 4v18.667c0 2.208 1.792 4 4 4h18.667c2.208 0 4-1.792 4-4v-18.667zM26.667 8v18.667c0 0.736-0.597 1.333-1.333 1.333 0 0-18.667 0-18.667 0-0.736 0-1.333-0.597-1.333-1.333 0 0 0-18.667 0-18.667 0-0.736 0.597-1.333 1.333-1.333 0 0 18.667 0 18.667 0 0.736 0 1.333 0.597 1.333 1.333z\\"></path>
          <path d=\\"M20 2.667v5.333c0 0.736 0.597 1.333 1.333 1.333s1.333-0.597 1.333-1.333v-5.333c0-0.736-0.597-1.333-1.333-1.333s-1.333 0.597-1.333 1.333z\\"></path>
          <path d=\\"M9.333 2.667v5.333c0 0.736 0.597 1.333 1.333 1.333s1.333-0.597 1.333-1.333v-5.333c0-0.736-0.597-1.333-1.333-1.333s-1.333 0.597-1.333 1.333z\\"></path>
          <path d=\\"M4 14.667h24c0.736 0 1.333-0.597 1.333-1.333s-0.597-1.333-1.333-1.333h-24c-0.736 0-1.333 0.597-1.333 1.333s0.597 1.333 1.333 1.333z\\"></path>
        </svg>
        <!---->
        <!---->
      </div>
    </div>
    <!---->
  </div>
  <!--v-if-->
</div>"
`;
